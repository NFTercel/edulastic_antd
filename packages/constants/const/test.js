module.exports = {
  type: {
    ASSESSMENT: "assessment",
    PRACTICE: "practice"
  },
  settingCategories: [
    { id: "mark-as-done", title: "MARK AS DONE" },
    { id: "release-scores", title: "RELEASE SCORES AUTOMATICALLY" },
    { id: "maximum-attempts-allowed", title: "MAXIMUM ATTEMPTS ALLOWED" },
    { id: "test-type", title: "TEST TYPE" },
    { id: "require-safe-exame-browser", title: "REQUIRE SAFE EXAME BROWSER" },
    { id: "show-questions", title: "RELEASE ANSWERS WITH GRADES" },
    { id: "suffle-question", title: "SUFFLE QUESTION" },
    { id: "show-answer-choice", title: "SHOW ANSWER CHOICE" },
    { id: "show-calculator", title: "SHOW CALCULATOR" },
    { id: "answer-on-paper", title: "ANSWER ON PAPER" },
    { id: "require-password", title: "REQUIRE PASSWORD" },
    { id: "evaluation-method", title: "EVALUATION METHOD" },
    { id: "performance-bands", title: "PERFORMANCE BANDS" },
    { id: "title", title: "TITLE" },
    { id: "navigations", title: "NAVIGATIONS / CONTROL" },
    { id: "accessibility", title: "ACCESSIBILITY" },
    { id: "ui-time", title: "UI / TIME" },
    { id: "administration", title: "ADMINISTRATION" }
  ],
  performanceBandsData: {
    ADVANCED: {
      bands: "Advanced",
      from: "100%"
    },
    MASTERY: {
      bands: "Mastery",
      from: "100%"
    },
    BASIC: {
      bands: "Basic",
      from: "100%"
    },
    APPROACHING_BASIC: {
      bands: "Approaching Basic",
      from: "100%"
    },
    UNSATISFACTORY: {
      bands: "Unsatisfactory",
      from: "100%"
    }
  },
  navigations: [
    "Intro Item",
    "Outro Item",
    "Previous",
    "Next",
    "Pause",
    "Save",
    "Submit",
    "Fullscreen",
    "Response Masking",
    "TOC Item Count",
    "Calculator",
    "Submit Criteria",
    "Warning if question not attempted",
    "Confirmation windows on submit",
    "Scroll to test element on test start",
    "Scroll to top on item change",
    "Exit Secure Browser",
    "Acknowledgements",
    "Table of Contents"
  ],
  completionTypes: { AUTOMATICALLY: "Automatically", MANUALLY: "Manually" },
  releaseGradeTypes: {
    DONT_RELEASE: "Do not release scores or responses",
    SCORE_ONLY: "Release scores only",
    WITH_RESPONSE: "Release scores and student responses",
    WITH_ANSWERS: "Release scores,student responses and correct answers"
  },
  releaseGradeKeys: ["DONT_RELEASE", "SCORE_ONLY", "WITH_RESPONSE", "WITH_ANSWERS"],
  calculators: {
    NONE: "None",
    SCIENTIFIC: "Scientific",
    BASIC: "Basic",
    GRAPHING: "Graphing"
  },
  evalTypes: {
    ALL_OR_NOTHING: "All or Nothing",
    PARTIAL_CREDIT: "Partial Credit"
  },
  accessibilities: {
    SHOW_COLOUR_SHCEME: "Show Colour Shceme",
    SHOW_FONT_SIZE: "Show Font Size",
    SHOW_ZOOM: "Show Zoom"
  }
};
