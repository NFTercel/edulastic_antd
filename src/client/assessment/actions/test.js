import { LOAD_TEST, FINISH_TEST, LOAD_PREVIOUS_RESPONSES } from "../constants/actions";

export const loadTestAction = payload => ({
  type: LOAD_TEST,
  payload
});

export const finishTestAcitivityAction = () => ({
  type: FINISH_TEST
});

export const loadPreviousReponseAction = () => ({
  type: LOAD_PREVIOUS_RESPONSES
});
