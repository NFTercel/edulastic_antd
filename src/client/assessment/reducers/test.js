import {
  LOAD_TEST_ITEMS,
  GOTO_ITEM,
  SET_TEST_ACTIVITY_ID,
  SET_TEST_ID,
  RESET_CURRENT_TEST_ITEM,
  SET_RESUME_STATUS,
  SET_TEST_LOADING_STATUS
} from "../constants/actions";

const initialState = {
  testActivityId: "",
  resume: false, // resume from last attempted?
  items: [],
  currentItem: 0,
  title: "",
  loading: true
};

const test = (state = initialState, { payload, type }) => {
  switch (type) {
    case LOAD_TEST_ITEMS:
      return {
        ...state,
        items: payload.items,
        title: payload.title,
        annotations: payload.annotations,
        docUrl: payload.docUrl
      };

    case GOTO_ITEM:
      return {
        ...state,
        currentItem: payload.item
      };

    case SET_TEST_ID:
      return {
        ...state,
        testId: payload.testId
      };

    case RESET_CURRENT_TEST_ITEM:
      return {
        ...state,
        currentItem: 0
      };
    case SET_TEST_ACTIVITY_ID:
      return {
        ...state,
        testActivityId: payload.testActivityId
      };
    case SET_RESUME_STATUS:
      return {
        ...state,
        resume: payload
      };
    case SET_TEST_LOADING_STATUS:
      return {
        ...state,
        loading: payload
      };
    default:
      return state;
  }
};

export default test;
