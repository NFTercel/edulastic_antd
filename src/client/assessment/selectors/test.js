import { createSelector } from "reselect";
import { cloneDeep } from "lodash";
import { getAnswersListSelector } from "./answers";

const stateSelector = state => state.test;

export const currentItemIndexSelector = createSelector(
  stateSelector,
  state => state.currentItem
);

export const itemsSelector = createSelector(
  stateSelector,
  state => state.items
);

export const currentItemSelector = createSelector(
  itemsSelector,
  currentItemIndexSelector,
  (items, index) => items[index]
);

export const currentQuestions = createSelector(
  currentItemSelector,
  item => {
    const rows = Array.isArray(item.rows) ? item.rows : [];
    return rows.reduce((acc, row) => {
      const widgets = Array.isArray(row.widgets) ? row.widgets : [];
      return [...acc, ...widgets];
    }, []);
  }
);

export const answersForCheck = createSelector(
  getAnswersListSelector,
  currentQuestions,
  answers => {
    const newAnswers = cloneDeep(answers);

    return newAnswers;
  }
);

export const itemQuestionsSelector = createSelector(
  currentItemSelector,
  item => {
    const questions = [];
    item.rows.forEach(row => {
      row.widgets.forEach(widget => {
        const qid = widget.reference;
        if (qid) questions.push(qid);
      });
    });
    return questions;
  }
);

export const testLoadingSelector = createSelector(
  stateSelector,
  state => state.loading
);
