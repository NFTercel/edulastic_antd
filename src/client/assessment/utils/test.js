import { keyBy as _keyBy, shuffle as _shuffle } from "lodash";

const shuffleOptions = (questions = [], qActivitiesById = {}) => {
  const optionsOrder = {};

  const modifiedQuestions = questions.map(question => {
    if (question.type === "multipleChoice") {
      let { shuffledOptions } = qActivitiesById[question.id] || {};
      if (!shuffledOptions) {
        shuffledOptions = _shuffle(question.options.map(item => item.value));
      }
      optionsOrder[question.id] = shuffledOptions;
      const optionByValue = _keyBy(question.options, "value");
      question.options = shuffledOptions.map(id => optionByValue[id]);
    }

    return question;
  });

  return [modifiedQuestions, optionsOrder];
};

export const ShuffleChoices = (testItems, questionActivities) => {
  const qActivitiesById = _keyBy(questionActivities, "qid");
  let shuffles = {};
  testItems.forEach(item => {
    let optionsOrder;
    [item.data.questions, optionsOrder] = shuffleOptions(item.data.questions, qActivitiesById);
    shuffles = {
      ...shuffles,
      ...optionsOrder
    };
  });

  return [testItems, shuffles];
};
