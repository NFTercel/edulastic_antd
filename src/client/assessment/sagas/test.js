import { testActivityApi, testsApi } from "@edulastic/api";
import { takeEvery, call, all, put, select } from "redux-saga/effects";
import { push } from "react-router-redux";
import { keyBy as _keyBy } from "lodash";

import {
  LOAD_TEST,
  LOAD_TEST_ITEMS,
  SET_TEST_ID,
  FINISH_TEST,
  LOAD_PREVIOUS_RESPONSES,
  LOAD_ANSWERS,
  SET_TEST_ACTIVITY_ID,
  LOAD_SCRATCH_PAD
} from "../constants/actions";
import { loadQuestionsAction } from "../actions/questions";
import { SET_RESUME_STATUS } from "../../student/Assignments/ducks";

const getQuestions = (testItems = []) => {
  const allQuestions = [];

  testItems.forEach(item => {
    const { questions = [] } = item.data;
    allQuestions.push(...questions);
  });

  return allQuestions;
};

function* loadTest({ payload }) {
  try {
    const { testActivityId, testId } = payload;
    yield put({
      type: SET_TEST_ID,
      payload: {
        testId
      }
    });

    // if testActivityId is passed, need to load previous responses as well!
    const getTestActivity = testActivityId ? call(testActivityApi.getById, testActivityId) : false;
    const [test, testActivity] = yield all([
      call(testsApi.getById, testId, {
        validation: true,
        data: true
      }),
      getTestActivity
    ]);

    const questions = getQuestions(test.testItems);
    yield put(loadQuestionsAction(_keyBy(questions, "id")));
    yield put({
      type: LOAD_TEST_ITEMS,
      payload: {
        items: test.testItems,
        title: test.title
      }
    });

    // if testActivity is present.
    if (testActivity) {
      let allAnswers = {};
      const userWork = {};

      yield put({
        type: SET_TEST_ACTIVITY_ID,
        payload: { testActivityId }
      });

      const { questionActivities } = testActivity;
      let lastAttemptedQuestion = questionActivities[0];

      questionActivities.forEach(item => {
        allAnswers = {
          ...allAnswers,
          [item.qid]: item.userResponse
        };
        if (item.updatedAt > lastAttemptedQuestion.updatedAt) {
          lastAttemptedQuestion = item;
        }
      });

      // get currentItem index;
      let lastAttendedQuestion = 0;
      if (lastAttemptedQuestion.testItemId) {
        test.testItems.forEach((item, index) => {
          if (item._id === lastAttemptedQuestion.testItemId) {
            lastAttendedQuestion = index;
          }
        });
      }

      // load previous responses
      yield put({
        type: LOAD_ANSWERS,
        payload: allAnswers
      });

      yield put({
        type: LOAD_SCRATCH_PAD,
        payload: userWork
      });

      // only load from previous attempted if resuming from assignments page
      const loadFromLast = yield select(state => state.test && state.test.resume);

      // move to last attended question
      if (loadFromLast) {
        yield put(push(`${lastAttendedQuestion}`));
        yield put({
          type: SET_RESUME_STATUS,
          payload: false
        });
      }
    }
  } catch (err) {
    console.error(err);
  }
}

// load users previous responses for a particular test
function* loadPreviousResponses() {
  try {
    const testActivityId = yield select(state => state.test && state.test.testActivityId);
    const answers = yield testActivityApi.previousResponses(testActivityId);
    yield put({
      type: LOAD_ANSWERS,
      payload: { ...answers }
    });
  } catch (err) {
    console.log(err);
  }
}

function* submitTest() {
  try {
    const testActivityId = yield select(state => state.test && state.test.testActivityId);
    if (testActivityId === "test") {
      console.log("practice test");
      return;
    }
    yield testActivityApi.submit(testActivityId);
    yield put(push("/home/reports"));
  } catch (err) {
    console.log(err);
  }
}

export default function* watcherSaga() {
  yield all([
    yield takeEvery(LOAD_TEST, loadTest),
    yield takeEvery(FINISH_TEST, submitTest),
    yield takeEvery(LOAD_PREVIOUS_RESPONSES, loadPreviousResponses)
  ]);
}
