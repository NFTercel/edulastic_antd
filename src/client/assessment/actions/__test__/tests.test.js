import { loadTestAction, finishTestAcitivityAction, loadPreviousReponseAction } from "../test";

describe("Test", () => {
  it("load test should return an action", () => {
    expect(loadTestAction({ testId: "6te42hgfd67821fa" })).toMatchSnapshot();
  });

  it("finish test activity should return an action", () => {
    expect(finishTestAcitivityAction({})).toMatchSnapshot();
  });

  it("load previous response should return an action", () => {
    expect(loadPreviousReponseAction({})).toMatchSnapshot();
  });
});
