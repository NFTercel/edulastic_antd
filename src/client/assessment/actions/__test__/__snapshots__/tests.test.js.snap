// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test finish test activity should return an action 1`] = `
Object {
  "type": "[test] finish test",
}
`;

exports[`Test load previous response should return an action 1`] = `
Object {
  "type": "[test] load previouls responses",
}
`;

exports[`Test load test should return an action 1`] = `
Object {
  "payload": Object {
    "test": Object {
      "testId": "6te42hgfd67821fa",
    },
    "testId": undefined,
  },
  "type": "[test] load test",
}
`;
